src/easylink/step.py

```diff
@@ -601,19 +493,162 @@ class HierarchicalStep(Step):
+        super().__init__(
+            step_name,
+            name,
+            input_slots,
+            output_slots,
+            input_slot_mappings,
+            output_slot_mappings,
+        )
+        self.nodes = nodes
+        """All sub-nodes (i.e. sub-``Steps``) that make up this ``HierarchicalStep``."""
+        for node in self.nodes:
+            node.set_parent_step(self)
+        self.edges = edges
+        """The :class:`~easylink.graph_components.EdgeParams` of the sub-nodes."""
+        self.step_graph = self._get_step_graph(nodes, edges)
+        """The :class:`~easylink.graph_components.StepGraph` i.e. the directed acyclic 
+        graph (DAG) of sub-nodes and their edges that make up this ``HierarchicalStep``."""
+        self.user_configurable = user_configurable
+        """Whether or not the ``HierarchicalStep`` is user-configurable. It is a convenience
+        attribute to allow for back-end ``HierarchicalStep`` creation that are not
+        user-facing (i.e. they do not need to provide a 'substeps' configuration key)."""
+
     @property
     def config_key(self):
         """The pipeline specification key required for a ``HierarchicalStep``."""
         return "substeps"
 
+    def validate_step(
+        self,
+        parent_config: LayeredConfigTree,
+        combined_implementations: LayeredConfigTree,
+        input_data_config: LayeredConfigTree,
+    ) -> dict[str, list[str]]:
+        """Validates the ``HierarchicalStep``.
+
+        Parameters
+        ----------
+        parent_config
+            The parent configuration of this ``HierarchicalStep``, i.e. includes
+            the step name in the outer keys.
+        combined_implementations
+            The configuration for any implementations to be combined.
+        input_data_config
+            The input data configuration for the entire pipeline.
+
+        Returns
+        -------
+            A dictionary of errors, where the keys are the ``HierarchicalStep``
+            name and the values are lists of error messages associated with the
+            given ``HierarchicalStep``.
+
+        Notes
+        -----
+        A ``HierarchicalStep`` can be in either a "leaf" or a "non-leaf" configuration
+        state and the validation process is different for each.
+
+        If the ``HierarchicalStep`` does not validate (i.e. errors are found and
+        the returned dictionary is non-empty), the tool will exit and the pipeline
+        will not run.
+
+        We attempt to batch error messages as much as possible, but there may be
+        times where the configuration is so ill-formed that we are unable to handle
+        all issues in one pass. In these cases, new errors may be found after the
+        initial ones are handled.
+        """
+        step_config = parent_config
+        if self.user_configurable:
+            if not self.name in step_config:
+                return {f"step {self.name}": ["The step is not configured."]}
```

Comment:
This seems maybe out of scope for this PR, but the code is here, and I don't think this is correct: the `self.name` part should be handled by the _parent_ step and unrelated to the `user_configurable` flag. 🤔

Reply:
I think the parent_config -> step_config handles this comment?

Reply:
Yes, it does! The `user_configurable` logic seems to be in the right place now.